internal/mode/static/state/graph/route_common.go

```diff
@@ -361,53 +362,71 @@ func bindRoutesToListeners(
+type hostPort struct {
+	hostname string
+	port     v1.PortNumber
+}
+
+func getListenerHostPortMap(listeners []*Listener) map[string]hostPort {
+	listenerHostPortMap := make(map[string]hostPort, len(listeners))
 	for _, l := range listeners {
-		listenerHostnameMap[l.Name] = getHostname(l.Source.Hostname)
+		listenerHostPortMap[l.Name] = hostPort{
+			hostname: getHostname(l.Source.Hostname),
+			port:     l.Source.Port,
+		}
 	}
+	return listenerHostPortMap
+}
 
+// isolateL7RouteListeners ensures listener isolation for all L7Routes.
+func isolateL7RouteListeners(routes []*L7Route, listenerHostPortMap map[string]hostPort) {
+	isL4Route := false
 	for _, route := range routes {
-		isolateHostnamesForParentRefs(route.ParentRefs, listenerHostnameMap)
+		isolateHostnamesForParentRefs(route.ParentRefs, listenerHostPortMap, isL4Route)
 	}
 }
 
 // isolateL4RouteListeners ensures listener isolation for all L4Routes.
-func isolateL4RouteListeners(routes []*L4Route, listeners []*Listener) {
-	listenerHostnameMap := make(map[string]string, len(listeners))
-	for _, l := range listeners {
-		listenerHostnameMap[l.Name] = getHostname(l.Source.Hostname)
-	}
-
+func isolateL4RouteListeners(routes []*L4Route, listenerHostPortMap map[string]hostPort) {
+	isL4Route := true
 	for _, route := range routes {
-		isolateHostnamesForParentRefs(route.ParentRefs, listenerHostnameMap)
+		isolateHostnamesForParentRefs(route.ParentRefs, listenerHostPortMap, isL4Route)
 	}
 }
 
 // isolateHostnamesForParentRefs iterates through the parentRefs of a route to identify the list of accepted hostnames
-// for each listener. If any accepted hostname belongs to another listener,
+// for each listener. If any accepted hostname belongs to another listener with the same port, then
 // it removes those hostnames to ensure listener isolation.
-func isolateHostnamesForParentRefs(parentRef []ParentRef, listenerHostnameMap map[string]string) {
+func isolateHostnamesForParentRefs(parentRef []ParentRef, listenerHostnameMap map[string]hostPort, isL4Route bool) {
 	for _, ref := range parentRef {
-		acceptedHostnames := ref.Attachment.AcceptedHostnames
+		// when sectionName is nil we allow all listeners to attach to the route
+		if ref.SectionName == nil {
+			continue
+		}
 
+		acceptedHostnames := ref.Attachment.AcceptedHostnames
 		hostnamesToRemoves := make(map[string]struct{})
 		for listenerName, hostnames := range acceptedHostnames {
 			if len(hostnames) == 0 {
 				continue
 			}
 			for _, h := range hostnames {
-				for lName, lHostname := range listenerHostnameMap {
+				for lName, lHostPort := range listenerHostnameMap {
 					// skip comparison if it is a catch all listener block
-					if lHostname == "" {
+					if lHostPort.hostname == "" {
 						continue
 					}
-					if h == lHostname && listenerName != lName {
-						hostnamesToRemoves[h] = struct{}{}
+
+					// for L4Routes, we only compare the hostname and listener name combination
```

Comment:
This comment should be updated since we are considering both types of routes in these conditionals.